sys/crypto/modes/cbc.c

```diff
@@ -47,23 +54,78 @@ int cipher_encrypt_cbc(const cipher_t *cipher, uint8_t iv[16],
 
         output_block_last = output + offset;
         offset += block_size;
-    } while (offset < length);
+    } while (offset < input_len);
 
     return offset;
 }
 
+int cipher_encrypt_cbc_pkcs7(const cipher_t *cipher, uint8_t iv[16],
+                                    const uint8_t *input, size_t input_len,
+                                    uint8_t *output, size_t output_buffer_size)
+{
+    uint8_t block_size = cipher->interface->block_size;
+
+    /*
+     * Output buffer size must be a multiple of block_size and large enough
+     * for the cipher text including the paddíng
+     */
+    if (output_buffer_size < (input_len + block_size - (input_len % block_size))) {
+        return CIPHER_ERR_BUFFER_TOO_SMALL;
+    }
+
+    /* Encrypt the bulk of data without the last block */
+    size_t padding_len_required = block_size - (input_len % block_size);
+    size_t last_block_offset = input_len - (input_len % block_size);
+
+    int encrypted_bytes = cipher_encrypt_cbc(cipher, iv, input, last_block_offset, output);
+    if (encrypted_bytes < 0) {
+        return encrypted_bytes;
+    }
+
+    /* Move the iv to the last block of the encrypted data */
+    if (encrypted_bytes > 0) {
+        iv = output + last_block_offset - cipher->interface->block_size;
+    }
+
+    /* Pad the final block and encrypt it */
+    uint8_t padded_data[block_size];
```

Comment:
Would this also work?

```suggestion
    uint8_t padded_data[CIPHER_MAX_CONTEXT_SIZE];
```

Reply:
I don't think `CIPHER_MAX_CONTEXT_SIZE` makes much sense, as it resolves to `CIPHERS_MAX_KEY_SIZE`. 
Do you actually mean `CIPHER_MAX_BLOCK_SIZE`?

Reply:
> Do you actually mean `CIPHER_MAX_BLOCK_SIZE`? That could work.

Yes if we know the max size it would be preferable over the VLA.

Reply:
As far as I can see, the block ciphers we have here all use the same block size (16 Bytes). It might become a problem if we ever implement and use block ciphers with different sizes, but that can be modified again later.